### ISSUE TITLE
Incorrect return value for twisted.persisted.styles._pickleFunction

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8504]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)| @markrwilliams reported|
|-|-|
|Trac ID|trac#8504|
|Type|defect|
|Created|2016-06-26 07:20:30Z|

`twisted.persisted.styles` attempts to register pickling and unpickling functions for Python functions (`FunctionType`).  T`_pickleFunction` returns `None` when asked to pickle a lambda because of [#7989](https://github.com/twisted/twisted/issues/7989).  But there are differences between PyPy's `pickle` implementation and CPython's so that returning `None` in PyPy results in a `TypeError` instead of the expected `PicklingError`. 

This is why `twisted.persisted.test.test_styles.UnpickleMethodTests.test_lambdaRaisesPicklingError` errors on PyPy.

`_pickleFunction` should not return `None`.  It should instead raise `PicklingError` (or a subclass of `pickle.PicklingError` and `cPickle.PicklingError` on CPython 2.7) when given a lambda.

Attachments:

* [pickle_problems.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2d3/2d3730e6849fab2a79e5421d3bf41cc86cf60f4a/d6aaa0337e732925b78ef9d39c4a4f424dc067bb.py) (1326 bytes) - added by markrwilliams on 2016-06-26 07:31:14Z - pickle_problems.py

<details><summary>Searchable metadata</summary>

```
trac-id__8504 8504
type__defect defect
reporter__markrwilliams markrwilliams
priority__normal normal
milestone__PyPy_support PyPy-support
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__pypy__persisted__pickle pypy, persisted, pickle
time__1466925630982654 1466925630982654
changetime__1474198049187875 1474198049187875
version__None None
owner__markrwilliams markrwilliams

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_1"></a>|@markrwilliams commented|
|-|-|

The ticket describes the minimal change necessary to get the persisted tests running meaningfully on PyPy.  I'd rather not skip them entirely because [#7989](https://github.com/twisted/twisted/issues/7989) shows that, for better or worse, Twisted users rely on persisted.

`twisted.persisted.styles` needs serious overhaul and the `TypeError` in PyPy is probably not a bug.  Both are true for the same fundamental reason: `pickle` is completely insane.

A custom pickle function for `FunctionType` is called only if: you are using cPickle on CPython 2.x-3.3 _or_ you are using PyPy _and_ you are attempting to serialize an instance of FunctionType that cPickle does not know how to serialize.

Here's why things are the way they are:

`copy_reg` does not affect the way the pure-Python `pickle` module (de)serializes built-in types in Python 2.  That's because the `Pickler` class always checks its own `dispatch` dictionary before `copy_reg.dispatch_table` (https://github.com/python/cpython/blob/2.7/Lib/pickle.py#L284-L290).  So the pure-Python `pickle` never calls `t.persisted.styles._pickleFunction` because `Pickler.dispatch` has its own take on functions (https://github.com/python/cpython/blob/2.7/Lib/pickle.py#L777).  This appears to have been true since at least Python 2.0 (https://github.com/python/cpython/blob/2.0/Lib/pickle.py)

The `cPickle/_pickle` module, however, _does_ respect `copyreg.dispatch_table` but _only_ for `FunctionType` and _only_ if its default pickling mechanism can't pickle the function. (https://github.com/python/cpython/blob/2.7/Modules/cPickle.c#L2725-L2735 & https://github.com/python/cpython/blob/2.7/Modules/cPickle.c#L2755-L2764).

Furthermore, if the `reduce` function returns something other than a string or tuple, a `PickleError` gets raised (https://github.com/python/cpython/blob/2.7/Modules/cPickle.c#L2806-L2816).

So `cPickle` (or `_pickle`) _do_ call `twisted.persisted.styles._pickleFunction`, and lambdas _do_ raise a `PicklingError`, but only because of _two_ failed conditions in the C code.  This is not good or robust.

It gets better.  The described behavior was true of `cPickle` up to and including Python 3.3  (https://github.com/python/cpython/blob/3.3/Modules/_pickle.c#L3221).

Then this happened before the 3.4 release:

https://github.com/python/cpython/commit/d38a7eee79e4de7817c958d183cf7b7809744276

Python 3.4 and above consequently always ignore `copyreg.dispatch_table` for built-in types, including `FunctionType`.

Now PyPy doesn't _have_ `cPickle`, and they seem to have erred on the side of making `pickle` respect `copy_reg.dispatch_table` (https://bitbucket.org/pypy/pypy/src/a0105e0d00dbd0f73d06fc704db704868a6c6ed2/lib-python/2.7/pickle.py?at=default&fileviewer=file-view-default#pickle.py-751).  That seems reasonable.  They don't expect a `reduce` function for `FunctionType` to return `None`, which also seems reasonable.  Thus the `TypeError` on `*None` (https://bitbucket.org/pypy/pypy/src/a0105e0d00dbd0f73d06fc704db704868a6c6ed2/lib-python/2.7/pickle.py?at=default&fileviewer=file-view-default#pickle.py-771).

This is all nuts.  The best thing would be to deprecate styles.  But until then, this ticket describes how to get tests exercising user-facing code paths on PyPy.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_2"></a>|@markrwilliams commented|
|-|-|

The attached Python program (`pickle_problems.py`) demonstrates the circumstances under which a `FunctionType` pickling function could ever be called.

Additionally, it makes sense that the various `pickle`s would try to special case at least `FunctionType`; if they didn't, a pickling function that returned an unpickling function would recur until the call stack overflowed, as `pickle` would try to serialize that function using the same pickling function.  To see this, change `_ignored` in `pickle_problems.py` to a lambda.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_3"></a>|@rodrigc set owner to @markrwilliams|
|-|-|

Based on review feedback from glyph in [ticket:8503] please re-work this patch
so that checks for _PY3 and _PYPY are not needed.  Something like:

```
import pickle
import cPickle
if cPickle.PicklingError is pickle.PicklingError:
   "..."
```

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_4"></a>|@markrwilliams commented|
|-|-|

Replying to [rodrigc](#note_3):
> Based on review feedback from glyph in [ticket:8503] please re-work this patch
> so that checks for _PY3 and _PYPY are not needed.  Something like:
> 
> ```
> import pickle
> import cPickle
> if cPickle.PicklingError is pickle.PicklingError:
>    "..."
> ```

I've re-opened https://github.com/twisted/twisted/pull/222 and refactored it to use this approach.

Please take a look when you get a chance!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|GitHub set status to `closed`|
|-|-|

In changeset 4df5aae23fb3945d6555d47280e09c85ae6e7fa2
```
#!CommitTicketReference repository="" revision="4df5aae23fb3945d6555d47280e09c85ae6e7fa2"
Merge pull request #222 from markrwilliams/pypy-persisted-styles-8504

Author: markrwilliams
Reviewer: rodrigc
Fixes: #8504

twisted.persisted.styles._pickleFunction explicitly raises a PicklingError when given a lambda.
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
